File: examples/vision-camera/modules/webrtc-source/ios/WebrtcSourceModule.swift
import RNFishjamClient
import ExpoModulesCore
import VisionCamera

public class WebrtcSourceModule: Module {
  static let registerPlugin: Void = {
    FrameProcessorPluginRegistry.addFrameProcessorPlugin("sendFrame") { proxy, options in
      return WebrtcFrameProcessorPlugin(proxy: proxy, options: options)
    }
  }()
  
  public required init(appContext: AppContext) {
    super.init(appContext: appContext)
    _ = WebrtcSourceModule.registerPlugin
  }
  
  public func definition() -> ModuleDefinition {
    Name("WebrtcSource")
    
    AsyncFunction("createVisionCameraTrack") {
      if let exisitingSource = WebrtcFrameProcessorPlugin.currentSource {
        RNFishjamProxy.remove(customSource: exisitingSource)
      }
      
      let source = WebrtcVisionCameraCustomSource()
      
      WebrtcFrameProcessorPlugin.currentSource = source
      
      try await RNFishjamProxy.add(customSource: source)
    }
    
    AsyncFunction("removeVisionCameraTrack") {
      if let exisitingSource = WebrtcFrameProcessorPlugin.currentSource {
        RNFishjamProxy.remove(customSource: exisitingSource)
      }
      WebrtcFrameProcessorPlugin.currentSource = nil
    }
  }
}

File: examples/vision-camera/modules/webrtc-source/android/src/main/java/io/fishjam/example/webrtcsource/WebrtcSourceModule.kt
package io.fishjam.example.webrtcsource

import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import io.fishjam.example.webrtcsource.webrtcframeprocessor.WebrtcFrameProcessorPlugin
import io.fishjam.reactnative.RNFishjamClient
import kotlinx.coroutines.runBlocking

class WebrtcSourceModule : Module() {
  // Packages needs to be initialized somewhere
  private val webrtcPlugin = WebrtcFrameProcessorPluginPackage()

  override fun definition() = ModuleDefinition {
    Name("WebrtcSource")

    OnDestroy {
      runBlocking {
        WebrtcFrameProcessorPlugin.currentSource?.let {
          RNFishjamClient.removeCustomSource(it)
        }
      }
    }

    AsyncFunction("createVisionCameraTrack") Coroutine { ->
      // Cleanup previous source if it wasn't removed for some reason
      WebrtcFrameProcessorPlugin.currentSource?.let {
        RNFishjamClient.removeCustomSource(it)
      }

      val source = WebrtcVisionCameraCustomSource()

      // Assign new source to the frame processor
      WebrtcFrameProcessorPlugin.currentSource = source

      // Create the source in fishjam (aka create track)
      RNFishjamClient.createCustomSource(source)
    }

    AsyncFunction("removeVisionCameraTrack") Coroutine { ->
      WebrtcFrameProcessorPlugin.currentSource?.let {
        RNFishjamClient.removeCustomSource(it)
      }
      WebrtcFrameProcessorPlugin.currentSource = null
    }
  }
}
